packages/api/app/index.js

```diff
@@ -62,6 +87,7 @@ app.use((err, _req, res, _next) => {
 const server = awsServerlessExpress.createServer(app);
 
 const handler = async (event, context) => {
+  await initEnvVars; // Wait for environment vars to resolve from initEnvVarsFunction
```

Comment:
Whoa. Does this pattern actually work in terms of forcing this to only be evaluated once per **lambda container** rather than once per **lambda invocation**?

Reply:
So the net effect should be one call to secrets manager per lambda container?

And then any API lambda interacting with the database would be making a call to secrets manager once per lambda invocation? Would that run afoul of the the API limits for secrets manager?

Reply:
`Whoa. Does this pattern actually work in terms of forcing this to only be evaluated once per lambda container rather than once per lambda invocation?`
`So the net effect should be one call to secrets manager per lambda container?`

To get the environment variables, yes.

Exceptions:
      There's  further optimization in how we're handling knex connections, as *each time* we get a knex config object we're making a call, and the API does make knex connections.     That is, however, in my opinion solidly out of scope
      
```
And then any API lambda interacting with the database would be making a call to secrets manager once per lambda invocation? Would that run afoul of the the API limits for secrets manager?
````

In terms of just the API lambda............ at some point, *yes*.     This creates a scaling issue given the SM API limit is 5k/second, so the practical limit will be if we're *warming* more than 5k lambdas per second (assuming no other SM access).   

If we want to be *sure*, it's reasonable to request proof before closing this ticket.    I can get started on that.

Reply:
Hmm yeah. I don't want to go overboard or blow up the scope, I've just become vary wary of API limits after the CMA experience. 

I guess we already are subject to the fact every lambda invocation makes a call to Secrets Manager, right? So we would already have a problem even without these changes if we had 5k lambdas per second.

But with these changes we could hit the same limit for some number of lambdas less than 5k lambdas per second. It's a tough call. Further changes here do seem out of scope but I'm worried about another scaling limit that only shows up once someone is hammering their system in production.

Thoughts?

Reply:
> I've just become vary wary of API limits after the CMA experience.

and more generally utilizing AWS APIs for nearly anything instance/workflow oriented, yes. 

I think this fundamentally comes down to an understanding of our requirements, and perhaps having a better understanding of the rate we pay the lambda warm-up cost.       

The 5k limit is feels pretty unlikely for current operational scenarios in a 'real' environment if we were only hitting the SM at lambda start up (and not, say, for database read/write) as it implies we're adding 5k new lambda container runtimes per second and doing so at a rate that a retry won't effectively throttle/remediate failures for short bursts. 

I'm actually far more concerned about how our current pattern has getKnexConfig/getKnexClient hit SM *at least once per lambda invocation*.    🤔

Reply:
Resolving this per the team discussion and will have a larger architecture meeting discussion on this topic in general